plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/DynamicFilteringJdbcSplitManager.java

```diff
@@ -0,0 +1,185 @@
+{
+    private static final Logger log = Logger.get(DynamicFilteringJdbcSplitManager.class);
+    private static final ConnectorSplitBatch EMPTY_BATCH = new ConnectorSplitBatch(ImmutableList.of(), false);
+
+    private final ConnectorSplitManager delegateSplitManager;
+    private final DynamicFilteringStats stats;
+
+    @Inject
+    public DynamicFilteringJdbcSplitManager(
+            @ForJdbcDynamicFiltering ConnectorSplitManager delegateSplitManager,
+            DynamicFilteringStats stats)
+    {
+        this.delegateSplitManager = requireNonNull(delegateSplitManager, "delegateSplitManager is null");
+        this.stats = requireNonNull(stats, "stats is null");
+    }
+
+    @Override
+    public ConnectorSplitSource getSplits(
+            ConnectorTransactionHandle transaction,
+            ConnectorSession session,
+            ConnectorTableHandle table,
+            DynamicFilter dynamicFilter,
+            Constraint constraint)
+    {
+        JdbcTableHandle tableHandle = (JdbcTableHandle) table;
+        // pushing DF through limit could reduce query performance
+        boolean hasLimit = tableHandle.getLimit().isPresent();
+        if (!dynamicFilteringEnabled(session) || hasLimit) {
+            return delegateSplitManager.getSplits(transaction, session, table, dynamicFilter, constraint);
+        }
+
+        return new DynamicFilteringSplitSource(transaction, session, (JdbcTableHandle) table, dynamicFilter, constraint);
+    }
+
+    private class DynamicFilteringSplitSource
+            implements ConnectorSplitSource
+    {
+        private final ConnectorTransactionHandle transaction;
+        private final ConnectorSession session;
+        private final JdbcTableHandle table;
+        private final DynamicFilter dynamicFilter;
+        private final Constraint constraint;
+        private final long dynamicFilteringTimeoutNanos;
+        private final long startNanos;
+
+        @GuardedBy("this")
+        private Optional<ConnectorSplitSource> delegateSplitSource = Optional.empty();
+
+        DynamicFilteringSplitSource(
+                ConnectorTransactionHandle transaction,
+                ConnectorSession session,
+                JdbcTableHandle table,
+                DynamicFilter dynamicFilter,
+                Constraint constraint)
+        {
+            this.transaction = requireNonNull(transaction, "transaction is null");
+            this.session = requireNonNull(session, "session is null");
+            this.table = requireNonNull(table, "table is null");
+            this.dynamicFilter = requireNonNull(dynamicFilter, "dynamicFilter is null");
+            this.constraint = requireNonNull(constraint, "constraint is null");
+            this.dynamicFilteringTimeoutNanos = (long) getDynamicFilteringWaitTimeout(session).getValue(NANOSECONDS);
+            this.startNanos = System.nanoTime();
+        }
+
+        @Override
+        public CompletableFuture<ConnectorSplitBatch> getNextBatch(int maxSize)
```

Comment:
I wonder if we could just wait in `JdbcRecordSetProvider` since DFs are sent back to workers? Do we compact DFs before sending them back to workers?